src/main/java/session/MenuSession.java

```diff
@@ -0,0 +1,224 @@
+            if(screen instanceof EntityScreen){
+                handleInput(currentSelection);
+            }
+        }
+    }
+
+    public MenuSession(SerializableMenuSession serializableMenuSession, RestoreService restoreService,
+                       InstallService installService) throws Exception {
+        this(serializableMenuSession.getUsername(), serializableMenuSession.getPassword(), serializableMenuSession.getDomain(),
+            serializableMenuSession.getInstallReference(), serializableMenuSession.getSerializedCommCareSession(), restoreService,
+                serializableMenuSession.getSessionId(), serializableMenuSession.getCurrentSelection(), installService);
+
+    }
+
+    private void restoreSessionFromStream(String serialiedCommCareSession) throws IOException, DeserializationException {
+        byte [] sessionBytes = Base64.decodeBase64(serialiedCommCareSession);
+        SessionFrame restoredFrame = new SessionFrame();
+        DataInputStream inputStream =
+                new DataInputStream(new ByteArrayInputStream(sessionBytes));
+        restoredFrame.readExternal(inputStream, new PrototypeFactory());
+        this.sessionWrapper.frame = restoredFrame;
+        this.sessionWrapper.syncState();
+    }
+
+    private String getSerializedCommCareSession() throws IOException {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        DataOutputStream serializedStream = new DataOutputStream(baos);
+        sessionWrapper.serializeSessionState(serializedStream);
+        String encoded = Base64.encodeBase64String(baos.toByteArray());
+        return encoded;
+    }
+
+    public SerializableMenuSession serialize() throws IOException {
+        SerializableMenuSession serializableMenuSession = new SerializableMenuSession();
+        serializableMenuSession.setUsername(this.username);
+        serializableMenuSession.setPassword(this.password);
+        serializableMenuSession.setDomain(this.domain);
+        serializableMenuSession.setInstallReference(this.installReference);
+        serializableMenuSession.setSessionId(this.sessionId);
+        serializableMenuSession.setSerializedCommCareSession(this.getSerializedCommCareSession());
+        serializableMenuSession.setCurrentSelection(this.currentSelection);
+        return serializableMenuSession;
+    }
+
+    public boolean handleInput(String input) throws CommCareSessionException {
+        boolean redrawing = screen.handleInputAndUpdateSession(sessionWrapper, input);
+        if(screen instanceof EntityScreen){
+            EntityScreen entityScreen = (EntityScreen) screen;
+            if(entityScreen.getCurrentScreen() instanceof EntityDetailSubscreen){
+                EntityDetailSubscreen entityDetailSubscreen = (EntityDetailSubscreen) entityScreen.getCurrentScreen();
+                int currentIndex = entityDetailSubscreen.getCurrentIndex();
+                currentSelection = input;
+            }
+        }
+        return redrawing;
```

Comment:
does redrawing seems kind of tertiary to what this function is doing. i feel like it could be 2 functions. one that gets redrawing and one that handles input (could be wrong though, you have more context)